ocaml/test/test_event.ml

```diff
@@ -0,0 +1,286 @@
+  let m = Mutex.create () in
+  let finished = ref false in
+  let pool = Db.Pool.get_all __context |> List.hd in
+  let key = "event_from_test" in
+  begin try Db.Pool.remove_from_other_config __context pool key with _ -> () end;
+  let (_: Thread.t) = Thread.create
+      (fun () ->
+         wait_for_pool_key __context key;
+         Mutex.execute m (fun () -> finished := true)
+      ) () in
+  Thread.delay 0.5;
+  Db.Pool.add_to_other_config __context pool key "1";
+  Thread.delay 0.5;
+  assert_bool "event_from_test" (Mutex.execute m (fun () -> !finished))
+
+let event_from_parallel_test () =
+  let __context, session_id = event_setup_common () in
+  let pool = Db.Pool.get_all __context |> List.hd in
+  let key = "event_next_test" in
+  begin try Db.Pool.remove_from_other_config __context pool key with _ -> () end;
+  let ok = ref true in
+  let (i_should_succeed: Thread.t) = Thread.create
+      (fun () ->
+         try
+           let _ = Xapi_event.from __context [] "" 2.0 in
+           () (* good *)
+         with e ->
+           Printf.printf "Caught unexpected error: %s\n" (ExnHelper.string_of_exn e);
+           ok := false;
+      ) () in
+  let (interfering_thread: Thread.t) = Thread.create
+      (fun () ->
+         wait_for_pool_key __context key
+      ) () in
+  Thread.delay 0.5; (* wait for both threads to block in Event.from *)
+  Db.Pool.add_to_other_config __context pool key "1";
+  Thread.join interfering_thread;
+  Thread.join i_should_succeed;
+  (* Check that Event.from didn't get cancelled by mistake *)
+  assert_bool "event_from_parallel_test" !ok
+
+let object_level_event_test session_id =
+  let __context, session_id = event_setup_common () in
+  let m = Mutex.create () in
+  let finished = ref false in
+  let reported_failure = ref false in
+  let vm_a = make_vm ~__context ~name_label:"vm_a" () in
+  let vm_b = make_vm ~__context ~name_label:"vm_b" () in
+
+  Printf.printf "watching %s\n%!" (Ref.string_of vm_a);
+  Printf.printf "ignoring %s\n%!" (Ref.string_of vm_b);
+
+  let key = "object_level_event_next" in
+
+  begin try Db.VM.remove_from_other_config __context vm_a key with _ -> () end;
+  begin try Db.VM.remove_from_other_config __context vm_b key with _ -> () end;
+
+  let (_: Thread.t) = Thread.create
+      (fun () ->
+         let token = ref "" in
+         while not (Mutex.execute m (fun () -> !finished)) do
+           Printf.printf "Calling event.from...\n%!";
+           let events = Xapi_event.from __context [ Printf.sprintf "vm/%s" (Ref.string_of vm_a) ] (!token) 10. |> parse_event_from in
+           Printf.printf "Got %d events\n%!" (List.length events.events);
+           List.iter
+             (fun event ->
+                if event.reference <> Ref.string_of vm_a then begin
+                  Printf.printf "event on %s which we aren't watching\n%!" event.reference;
+                  Mutex.execute m
+                    (fun () ->
+                       reported_failure := true;
```

Comment:
In quicktest, we also call the `failed` function at this point (and other points where we set `reported_failure` to true), to mark the test has having failed... but here we do nothing, just carry on, and so far as I can see the testcase will seem to succeed (except for the meaning of the words written to stdout).

Elsewhere in this new test code, we use `assert_bool` or similar as a replacement for `failed`, which seems to make more sense. See comment below...